internal/environment/environment.go

```diff
@@ -205,36 +206,10 @@ func (n *NomadEnvironment) Scale(apiClient nomad.ExecutorAPI) error {
 	if required > 0 {
 		return n.createRunners(apiClient, uint(required), true)
 	} else {
-		return n.removeRunners(apiClient, uint(-required))
+		return n.removeIdleRunners(apiClient, uint(-required))
```

Comment:
If we always delete the execution environment incl. all runners, do we need to keep the `else` path in the `Scale` method? Is there any use case we could decrease the scale without removing the environment before?

Reply:
One reason to leave it as it was would be that several e2e tests would otherwise be flaky and in need of revision.
These tests create the same execution environment several times in a short period of time. Sometimes it happens that a new runner was created from a previous run exactly in the period between the deletion and the scaling. In this case, idle runners have to be removed.
But maybe we should review the e2e tests rather than having unused code running in production!?

Reply:
I almost got it and see the issue. Just one question remaining: Who is responsible for these additional (unwanted) runners in the test scenario? Is it caused by an older, asynchronous request from Poseidon to Nomad which gets executed too "late" in Nomad (but has been planned successfully beforehand) or is it Poseidon requesting new runners in that moment? If it is the first case: Why can't we stop these allocations from spawning? If I see that correctly in Nomad, they already have an allocation ID while being pending (which would make them accessible for the delete request). And if it is the second case: Why is Poseidon still requesting these new runners? Is there some concurrent process that is not aware of the deletion?

I wondered whether this bug could also occur in our production environment (based on any of the two assumptions above). Sure, I am also fine with allowing to reduce the Prewarming pool size (instead of returning an error) but still wondered about this behavior.

Reply:
It's your first case. Let's take a look at an example, the e2e test `TestCreateOrUpdateEnvironment`.
- Line 41 the execution environment is created the first time. 
  - Poseidon registers the template job and one additional runner
- Line 50 the execution environment is updated
  - Poseidon updates the template job and registers one additional runner
- Line 59 the execution environment is updated
  - The execution environment is updated
    - It is deleted
    - Now the two runner could be allocated
    - When we now try to scale (and add one additional runner) we should remove one runner (instead)

This behavior can be seen in this log [flakyCase.txt](https://github.com/openHPI/poseidon/files/7739699/flakyCase.txt). The inconspicuous process, on the other hand, generates this log [nonFlakyCase.txt](https://github.com/openHPI/poseidon/files/7739701/nonFlakyCase.txt)

> Why can't we stop these allocations from spawning? If I see that correctly in Nomad, they already have an allocation ID while being pending (which would make them accessible for the delete request).

Just with the logs so far, it is noticeable that the template job for the created environment 42 was never started. By directly telling Nomad to delete the template job (line 211 environment.go), the spawning of the allocation was stopped.
The runners themselves, on the other hand, are only collected via the "LoadRunnerIDs" function. [In a test](https://github.com/openHPI/poseidon/files/7739725/whenIsARunnerIdFetchable.txt), I logged the "Fetched runner ids" until the runners could be listed. It could be determined that this is only the case when (/25ms before) Poseidon knows that the allocation has also been started.

This leads to the two most obvious options, either to cache the IDs of all created runners, or to accept the inaccuracy that when deleting an environment, pending runners are not deleted.

Reply:
Your log and the specific example is very helpful, thanks! Before going into detail, my main insight is: The tests are not independent of each other and (even worse) there are race conditions between those. I understand where this behavior is coming from but would argue that it is not desirable. I would rather tend to have the tests more independent (for example, in our RSpec tests suite the execution order of tests is randomized for each run forcing each test to be self-contained). While this is completely out of scope for this PR, I think these flaky tests unveiled an issue we should continue to work on.

Back to the example:
I expected that once line 50 gets executed, the previous runner (from line 41) gets deleted anyway so that only one runner is available. Based on your explanation and the log, that runner might not get deleted. It has been scheduled in Nomad but no allocationID is preserved in Poseidon's memory so that it is returned as part of the `LoadRunnerIDs` call (probably because of line 105 in nomad.go?)

To minimize erroneous behavior, I would tend to delete all allocations (even pending one) for the given execution environment. Besides the two options you mentioned (cache IDs, remove runners for outdated execution environments) I thought of adapting the `LoadRunnerIDs` to return all allocation IDs for an environment independent of their status (pending, running, [completed]). Do you think this could work?

Reply:
Yes, that is the best option and [it works](https://github.com/openHPI/poseidon/files/7762793/TestCreateOrUpdateEnvironment.log). Thank you :pray:

Reply:
I am glad it worked out, thank you! 💐